spec.md

```diff
@@ -0,0 +1,728 @@
+EEI specification
+=================
+
+Introduction
+------------
+
+This document aims to specify an ewasm VM in a way useful to contract writers and VM implementers.
+To this end, multiple things are specified:
+
+-   The extra state that a VM needs to have around to successfully respond to calls into the EEI.
+-   The EEI (Ethereum Environment Interface), currently specified loosley [here](eth_interface.md).
+
+### Terminology
+
+**EEI**: The *Ethereum Environment Interface* refers to the layer between the Ethereum Client code and the execution engine.
+         **TODO**: Is *EEI* the right term to use?
+
+**Ethereum Client**: Code which can interact with the blockchain (read/validate and sending transactions).
+
+**execution engine**: The underlying "hardware" of the VM, implementing the basic computational functions.
+
+**VM**: The VM is the combination of an Ethereum Client and the execution engine.
+
+### Notation
+
+We are using [K Framework] notation to specify the EEI, which makes this specification executable.
+
+```k
+requires "domains.k"
+
+module EEI
+    imports DOMAINS
+```
+
+Execution State
+---------------
+
+Below both a K rule and a prose description of each state transition is given.
+The state is specified using a K *configuration*.
+Each XML-like *cell* contains a field which is relevant to Ethereum client execution (eg. below the first cell is the `<k>` cell).
+The default/initial values of the cells are provided along with the declaration of the configuration.
+
+In the texual rules below, we'll refer to cells by accesing subcells with the `.` operator.
+For example, we would access contents of the `statusCode` cell with `eei.statusCode`.
+For cells that contain elements of K builtin sorts `Map`, `List`, and `Set`, we'll use standard K operators for referring to their contents.
+For example, we can access the third element of the `returnData` cell's list using `eei.call.returnData[2]`.
+
+For some cells, we have comments following the cell declarations with the name the [Yellow Paper] gives to that element of the state.
+
+```k
+    configuration
+      <k> $PGM:EthereumSimulation </k>
+      <eei>
+        <statusCode> .StatusCode </statusCode>
+```
+
+The `<callState>` sub-configuration can be saved/restored when needed between calls.
+
+```k
+        <call>
+          <callDepth>  0     </callDepth>
+          <returnData> .List </returnData>
+
+          <id>        0        </id>        // I_a
+          <program>   .Program </program>   // I_b
+          <caller>    0        </caller>    // I_s
+          <callData>  .List    </callData>  // I_d
+          <callValue> 0        </callValue> // I_v
+
+          <gas>        0 </gas>        // \mu_g
+          <memoryUsed> 0 </memoryUsed> // \mu_i
+        </call>
```

Comment:
I guess there should be `static` flag eventually? e.g we can't perform selfdestructing in a static call

Reply:
Yeah, I'll add it back in.

Reply:
Actually, I think if we add back in the `static` flag, should also add back in the actual checks (eg. on `EEI.setAccountStorage`). I think I'll do that as a second PR into the `eei-v2` branch after merging, because then can inspect it atomically.